lib/auth/auth_with_roles.go

```diff
@@ -1167,21 +1167,51 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 	var roles []string
 	var traits wrappers.Traits
 
-	switch {
-	case a.hasBuiltinRole(string(teleport.RoleAdmin)):
-		// If it's an admin generating the certificate, the roles and traits for
-		// the user have to be fetched from the backend. This should be safe since
-		// this is typically done against a local user.
-		user, err := a.GetUser(req.Username, false)
-		if err != nil {
-			return nil, trace.Wrap(err)
+	// this prevents clients who have no chance at getting a cert and impersonating anyone
+	// from enumerating local users and hitting database
+	if !a.hasBuiltinRole(string(teleport.RoleAdmin)) && !a.context.Checker.CanImpersonateSomeone() && req.Username != a.context.User.GetName() {
+		return nil, trace.AccessDenied("access denied: impersonation is not allowed")
+	}
+
+	// Prohibit recursive impersonation behavior:
+	//
+	// Alice can impersonate Bob
+	// Bob can impersonate Grace <- this code block prohibits the escape
+	//
+	// Allow cases:
+	//
+	// Alice can impersonate Bob
+	//
+	// Bob (impersonated by Alice) can renew the cert with route to cluster
+	//
+	if a.context.Identity != nil && a.context.Identity.GetIdentity().Impersonator != "" {
+		if len(req.AccessRequests) > 0 {
+			return nil, trace.AccessDenied("access denied: impersonating user can not request new roles")
```

Comment:
Nit: I think this should be "impersonated user" not "impersonating user".  The root identity is the one impersonating.  The identity being rejected here is being impersonated.